File: server/queueCoordinator.ts
import { WebSocket } from 'ws'
import type { ParticipantRole } from './sessionRouter.js'

interface SessionSlots {
  customer?: WebSocket
  consultant?: WebSocket
}

interface ParticipantInfo {
  sessionId: string
  role: ParticipantRole
}

interface JoinResultAccepted {
  accepted: true
}

interface JoinResultRejected {
  accepted: false
  reason: string
}

export type JoinResult = JoinResultAccepted | JoinResultRejected

export class QueueCoordinator {
  private readonly sessions = new Map<string, SessionSlots>()
  private readonly participants = new WeakMap<WebSocket, ParticipantInfo>()

  onJoin(socket: WebSocket, sessionId: string, role: ParticipantRole): JoinResult {
    this.onDisconnect(socket)

    const slots = this.sessions.get(sessionId) ?? {}

    if (role === 'customer' && slots.customer && slots.customer !== socket) {
      return { accepted: false, reason: 'Session already has a customer connected.' }
    }

    if (role === 'consultant' && slots.consultant && slots.consultant !== socket) {
      return { accepted: false, reason: 'Session already has a consultant connected.' }
    }

    if (role === 'customer') {
      slots.customer = socket
    } else {
      slots.consultant = socket
    }

    this.sessions.set(sessionId, slots)
    this.participants.set(socket, { sessionId, role })
    return { accepted: true }
  }

  onLeave(socket: WebSocket): void {
    this.onDisconnect(socket)
  }

  onDisconnect(socket: WebSocket): void {
    const info = this.participants.get(socket)
    this.participants.delete(socket)

    if (!info) {
      return
    }

    const slots = this.sessions.get(info.sessionId)
    if (!slots) {
      return
    }

    if (info.role === 'customer' && slots.customer === socket) {
      slots.customer = undefined
    }

    if (info.role === 'consultant' && slots.consultant === socket) {
      slots.consultant = undefined
    }

    if (!slots.customer && !slots.consultant) {
      this.sessions.delete(info.sessionId)
    }
  }

  getStats(): { activeSessions: number; waitingCustomers: number; waitingConsultants: number } {
    let waitingCustomers = 0
    let waitingConsultants = 0

    for (const slots of this.sessions.values()) {
      if (slots.customer && !slots.consultant) {
        waitingCustomers += 1
      }

      if (!slots.customer && slots.consultant) {
        waitingConsultants += 1
      }
    }

    return {
      activeSessions: this.sessions.size,
      waitingCustomers,
      waitingConsultants,
    }
  }
}

File: backend/src/main/java/com/videoshop/backend/signaling/QueueCoordinator.java
package com.videoshop.backend.signaling;

import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class QueueCoordinator {

	private final Map<String, SessionSlots> sessions = new ConcurrentHashMap<>();
	private final Map<String, ParticipantState> participants = new ConcurrentHashMap<>();

	public synchronized JoinResult onJoin(WebSocketSession session, String sessionId, ParticipantRole role) {
		onDisconnect(session);

		SessionSlots slots = sessions.getOrDefault(sessionId, new SessionSlots());

		if (role == ParticipantRole.CUSTOMER && slots.customer() != null && !slots.customer().equals(session)) {
			return JoinResult.reject("Session already has a customer connected.");
		}

		if (role == ParticipantRole.CONSULTANT && slots.consultant() != null && !slots.consultant().equals(session)) {
			return JoinResult.reject("Session already has a consultant connected.");
		}

		if (role == ParticipantRole.CUSTOMER) {
			slots = slots.withCustomer(session);
		} else {
			slots = slots.withConsultant(session);
		}

		sessions.put(sessionId, slots);
		participants.put(session.getId(), new ParticipantState(sessionId, role));
		return JoinResult.accept();
	}

	public synchronized void onLeave(WebSocketSession session) {
		onDisconnect(session);
	}

	public synchronized void onDisconnect(WebSocketSession session) {
		ParticipantState info = participants.remove(session.getId());
		if (info == null || info.sessionId() == null || info.role() == null) {
			return;
		}

		SessionSlots slots = sessions.get(info.sessionId());
		if (slots == null) {
			return;
		}

		if (info.role() == ParticipantRole.CUSTOMER && session.equals(slots.customer())) {
			slots = slots.withCustomer(null);
		}

		if (info.role() == ParticipantRole.CONSULTANT && session.equals(slots.consultant())) {
			slots = slots.withConsultant(null);
		}

		if (slots.customer() == null && slots.consultant() == null) {
			sessions.remove(info.sessionId());
			return;
		}

		sessions.put(info.sessionId(), slots);
	}

	public synchronized QueueStats getStats() {
		int waitingCustomers = 0;
		int waitingConsultants = 0;

		for (SessionSlots slots : sessions.values()) {
			if (slots.customer() != null && slots.consultant() == null) {
				waitingCustomers += 1;
			}

			if (slots.customer() == null && slots.consultant() != null) {
				waitingConsultants += 1;
			}
		}

		return new QueueStats(sessions.size(), waitingCustomers, waitingConsultants);
	}

	public record JoinResult(boolean accepted, String reason) {
		public static JoinResult accept() {
			return new JoinResult(true, null);
		}

		public static JoinResult reject(String reason) {
			return new JoinResult(false, reason);
		}
	}

	public record QueueStats(int activeSessions, int waitingCustomers, int waitingConsultants) {
	}

	private record SessionSlots(WebSocketSession customer, WebSocketSession consultant) {
		private SessionSlots() {
			this(null, null);
		}

		private SessionSlots withCustomer(WebSocketSession customer) {
			return new SessionSlots(customer, consultant);
		}

		private SessionSlots withConsultant(WebSocketSession consultant) {
			return new SessionSlots(customer, consultant);
		}
	}
}